File: backend/app/api/schemas/criteria.py
"""Request/response schemas for success criteria.

These mirror `app.domain.criteria`'s wire form field for field, because that
form is what is stored, what the API speaks, and what the generated frontend
types are derived from — three views of one document, so they have to be one
shape.

The split of responsibility matches `app.api.schemas.anchors`: the schema
owns the *structure* (which members the tagged union has, which fields each
carries, what type they are) and the domain owns the *rules* (fractions in
[0, 1], a band that is not inverted, a ceiling whose unit matches its
channel). Restating the numeric bounds here would mean two places to change
and two messages to keep in agreement, and the domain's message is the better
one.
"""

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.domain.anchors import AnchorType
from app.domain.criteria import (
    DEFAULT_BAND_SMOOTHING_S,
    DEFAULT_CEILING_SMOOTHING_S,
    CriterionKind,
    StepSelectorKind,
)
from app.domain.workout import Channel, ChannelUnit, StepRole


class StepSelectorSchema(BaseModel):
    """Which flattened steps a criterion applies to."""

    model_config = ConfigDict(extra="forbid")

    kind: StepSelectorKind
    #: Required for a ``role`` selector, absent otherwise.
    role: StepRole | None = None
    #: Required for an ``index`` selector, absent otherwise.
    index: int | None = None


class BandSchema(BaseModel):
    """An acceptable range around a step's prescribed target, as fractions."""

    model_config = ConfigDict(extra="forbid")

    channel: Channel
    #: Lower bound as a fraction of the prescribed target (``0.95`` = -5 %).
    low: float
    #: Upper bound as a fraction of the prescribed target.
    high: float
    #: Seconds of trailing rolling mean applied to the channel before it is
    #: compared to the band; 0 means raw samples. Omit to take the default —
    #: the window is part of the frozen intent, so it is always returned.
    smoothing_s: int = DEFAULT_BAND_SMOOTHING_S


class PercentLimitSchema(BaseModel):
    """A ceiling expressed as a fraction of an anchor."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal["percent_of_anchor"] = "percent_of_anchor"
    anchor_type: AnchorType
    pct: float


class AbsoluteLimitSchema(BaseModel):
    """A ceiling expressed in the channel's own unit."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal["absolute"] = "absolute"
    value: float
    unit: ChannelUnit


LimitSchema = Annotated[
    PercentLimitSchema | AbsoluteLimitSchema, Field(discriminator="kind")
]


class TimeInBandSchema(BaseModel):
    """At least ``min_fraction`` of the selected steps' time inside the band."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal[CriterionKind.TIME_IN_BAND] = CriterionKind.TIME_IN_BAND
    selector: StepSelectorSchema
    band: BandSchema
    min_fraction: float


class DurationFloorSchema(BaseModel):
    """The session must last at least this long."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal[CriterionKind.DURATION_FLOOR] = CriterionKind.DURATION_FLOOR
    min_seconds: int


class CeilingSchema(BaseModel):
    """No more than this long spent above this limit on this channel."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal[CriterionKind.CEILING] = CriterionKind.CEILING
    channel: Channel
    limit: LimitSchema
    max_seconds_above: int
    #: Seconds of trailing rolling mean applied before the comparison.
    #: Defaults to 0 — raw — because a ceiling is about excursions and
    #: smoothing hides them.
    smoothing_s: int = DEFAULT_CEILING_SMOOTHING_S


class SetsCompletedSchema(BaseModel):
    """At least this fraction of the prescribed strength sets performed."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal[CriterionKind.SETS_COMPLETED] = CriterionKind.SETS_COMPLETED
    min_fraction: float


class LoadWithinSchema(BaseModel):
    """Loads used within this relative tolerance of what was prescribed."""

    model_config = ConfigDict(extra="forbid")

    kind: Literal[CriterionKind.LOAD_WITHIN] = CriterionKind.LOAD_WITHIN
    pct_tolerance: float


#: One success criterion, discriminated by ``kind``.
SuccessCriterionSchema = Annotated[
    TimeInBandSchema
    | DurationFloorSchema
    | CeilingSchema
    | SetsCompletedSchema
    | LoadWithinSchema,
    Field(discriminator="kind"),
]

File: frontend/lib/criteria.ts
import type { components } from "@/generated/api/schema";

import { formatDurationClock, formatPercent } from "@/lib/format";

type Schemas = components["schemas"];

/** The tagged union of machine-checkable success criteria (build plan WP-2.7). */
export type SuccessCriterion =
  | Schemas["TimeInBandSchema"]
  | Schemas["DurationFloorSchema"]
  | Schemas["CeilingSchema"]
  | Schemas["SetsCompletedSchema"]
  | Schemas["LoadWithinSchema"];

const ANCHOR_LABELS: Record<Schemas["AnchorType"], string> = {
  ftp: "FTP",
  lthr: "LTHR",
  max_hr: "max HR",
  cp: "CP",
  w_prime: "W′",
};

const CHANNEL_LABELS: Record<Schemas["Channel"], string> = {
  power: "power",
  hr: "heart rate",
  cadence: "cadence",
};

const STEP_ROLE_LABELS: Record<Schemas["StepRole"], string> = {
  warmup: "warm-up",
  work: "work",
  recovery: "recovery",
  rest: "rest",
  cooldown: "cool-down",
};

/**
 * One criterion as a sentence an athlete can check themselves against.
 *
 * The wire format is a tagged union chosen to be *evaluable* — selectors,
 * fractions, anchor references — which makes it unreadable as-is. This is the
 * only place that translation lives, so the calendar sheet, the creator's
 * criteria editor (slice 2) and any later verdict screen all phrase a
 * criterion the same way.
 */
export function describeCriterion(criterion: SuccessCriterion): string {
  switch (criterion.kind) {
    case "time_in_band": {
      const channel = CHANNEL_LABELS[criterion.band.channel];
      const low = formatPercent(criterion.band.low);
      const high = formatPercent(criterion.band.high);
      return `${formatPercent(criterion.min_fraction)} of ${describeSelector(
        criterion.selector,
      )} within ${low}–${high} of the prescribed ${channel}`;
    }
    case "duration_floor":
      return `Lasts at least ${formatDurationClock(criterion.min_seconds)}`;
    case "ceiling": {
      const channel = CHANNEL_LABELS[criterion.channel];
      const limit =
        criterion.limit.kind === "percent_of_anchor"
          ? `${formatPercent(criterion.limit.pct)} of ${
              ANCHOR_LABELS[criterion.limit.anchor_type]
            }`
          : `${criterion.limit.value} ${criterion.limit.unit}`;
      return `No more than ${formatDurationClock(
        criterion.max_seconds_above,
      )} with ${channel} above ${limit}`;
    }
    case "sets_completed":
      return `${formatPercent(criterion.min_fraction)} of the prescribed sets completed`;
    case "load_within":
      return `Loads within ${formatPercent(criterion.pct_tolerance)} of what was prescribed`;
  }
}

/** The tag of each criterion in the union. */
export type CriterionKind = SuccessCriterion["kind"];

/** How the criteria editor names each kind in its "add" menu. */
export const CRITERION_KIND_LABELS: Readonly<Record<CriterionKind, string>> = {
  time_in_band: "Time in band",
  duration_floor: "Minimum duration",
  ceiling: "Ceiling",
  sets_completed: "Sets completed",
  load_within: "Load within tolerance",
};

/**
 * Which criteria a discipline can be judged by.
 *
 * Mirrors the domain's `STRENGTH_ONLY_KINDS` / `ENDURANCE_ONLY_KINDS`: a
 * `time_in_band` on a lifting session refers to a power trace that will never
 * exist, and the backend refuses it. Offering it and then failing would be a
 * worse way to say the same thing.
 */
export function criterionKindsFor(
  discipline: Schemas["Discipline"],
): CriterionKind[] {
  return discipline === "cycling"
    ? ["time_in_band", "duration_floor", "ceiling"]
    : ["sets_completed", "load_within", "duration_floor"];
}

/**
 * A criterion of this kind with defensible starting values.
 *
 * Not zeroes: a criterion is a rule, and a rule of "at least 0% of the time"
 * passes every session ever ridden. These are the values a coach would write
 * first and then adjust.
 */
export function blankCriterion(kind: CriterionKind): SuccessCriterion {
  switch (kind) {
    case "time_in_band":
      return {
        kind: "time_in_band",
        // 30 s is the conventional window for steady work, and the
        // default the backend applies when a band does not state one.
        band: { channel: "power", low: 0.95, high: 1.05, smoothing_s: 30 },
        min_fraction: 0.8,
        selector: { kind: "role", role: "work", index: null },
      };
    case "duration_floor":
      return { kind: "duration_floor", min_seconds: 3600 };
    case "ceiling":
      return {
        kind: "ceiling",
        channel: "hr",
        limit: { kind: "percent_of_anchor", anchor_type: "lthr", pct: 1.0 },
        max_seconds_above: 300,
        // Raw: a ceiling is about excursions, and smoothing hides them.
        smoothing_s: 0,
      };
    case "sets_completed":
      return { kind: "sets_completed", min_fraction: 0.9 };
    case "load_within":
      return { kind: "load_within", pct_tolerance: 0.05 };
  }
}

/** Which steps a criterion applies to, in words. */
export function describeSelector(
  selector: Schemas["StepSelectorSchema"],
): string {
  switch (selector.kind) {
    case "all":
      return "the session's time";
    case "role":
      return selector.role
        ? `the ${STEP_ROLE_LABELS[selector.role]} steps' time`
        : "the selected steps' time";
    case "index":
      return selector.index === null || selector.index === undefined
        ? "the selected step's time"
        : `step ${selector.index + 1}'s time`;
  }
}
